colossalai/nn/optimizer/distributed_adafactor.py

```diff
@@ -0,0 +1,364 @@
+        param_scale = 1.0
+        if param_group["scale_parameter"]:
+            param_scale = max(param_group["eps"][1], param_state["RMS"])
+        return param_scale * rel_step_sz
+
+    @staticmethod
+    def _get_options(param_group, param_shape):
+        """
+        Determines whether the current param is factored
+        Args:
+            param_group : param group
+            param_shape : Original Shape of param
+
+        """
+        factored = len(param_shape) >= 2
+        use_first_moment = param_group["beta1"] is not None
+        return factored, use_first_moment
+
+    @staticmethod
+    def _rms(tensor, param_is_dtensor, tp_size, dp_size, tp_group, dp_group):
+        tensor_sum = tensor.pow(2).sum()
+        num_of_element = tensor.numel()
+        # reduce sum on tp group if exist
+        if tp_size > 1 and param_is_dtensor:
+            dist.all_reduce(tensor_sum, group=tp_group)
+            num_of_element = num_of_element * tp_size
+        # reduce sum on dp group if exist
+        if dp_size > 1 and param_is_dtensor:
+            dist.all_reduce(tensor_sum, group=dp_group)
+            num_of_element = num_of_element * dp_size
+        # div num of element
+        rms = (tensor_sum / num_of_element).sqrt()
+        return rms
+
+    @staticmethod
+    def _approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col):
+        r_factor = (exp_avg_sq_row / exp_avg_sq_row.mean(dim=-1, keepdim=True)).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    # approx_sq_grad for row parallel weight
+    @staticmethod
+    def _approx_sq_grad_row_parallel(exp_avg_sq_row, exp_avg_sq_col, sq_row_meam):
+        # row_meam = sq_row_meam
+        r_factor = (exp_avg_sq_row / sq_row_meam).rsqrt_().unsqueeze(-1)
+        c_factor = exp_avg_sq_col.unsqueeze(-2).rsqrt()
+        return torch.mul(r_factor, c_factor)
+
+    def _col_parallel_factor(self, update, grad, state, grad_shape, beta2t):
+        update_reshape = update.view(-1, grad_shape[1])
+        grad_reshape = grad.view(-1, grad_shape[1])
+        exp_avg_sq_row = state["exp_avg_sq_row"]  # [H/dp]
+        exp_avg_sq_col = state["exp_avg_sq_col"]  # [W/tp]
+        exp_avg_sq_row.mul_(beta2t).add_(update_reshape.mean(dim=-1), alpha=(1.0 - beta2t))
+        exp_avg_sq_col.mul_(beta2t).add_(update_reshape.mean(dim=-2), alpha=(1.0 - beta2t))
+        dist.all_reduce(exp_avg_sq_row, group=self.tensor_parallel_group)
+        exp_avg_sq_row.div_(self.tensor_parallel_size)
+        update_reshape = self._approx_sq_grad(exp_avg_sq_row, exp_avg_sq_col)
+        update_reshape.mul_(grad_reshape)
+        if self.use_zero:
+            update = update_reshape.view(-1)
+        else:
+            update = update_reshape
+        return update
+
+    def _row_parallel_factor(self, update, grad, state, grad_shape, beta2t):
+        if grad_shape[0] % self.data_parallel_size != 0:
```

Comment:
You should check whether zero is used

Reply:
Fixed; Also, _row_parallel_factor and _col_parallel_factor cover all cases of use zero and tp. We omit the zero judgement (if you don't use zero, the dp size will be set to 1 by default, which doesn't happen for gather or reduce in a dp group), but detect the shape residuals that could be caused by zero.